File: src/core/format.ts
import type { LimitWindow } from "./types";

export function fmtTime(ms: number | null): string {
  if (ms === null) return "?";
  return new Date(ms).toLocaleTimeString([], { hour: "2-digit", minute: "2-digit" });
}

/** "4h 49m", "1d 21h", "12m" */
export function fmtDuration(ms: number): string {
  const mins = Math.max(0, Math.round(ms / 60_000));
  const d = Math.floor(mins / 1440);
  const h = Math.floor((mins % 1440) / 60);
  const m = mins % 60;
  if (d) return `${d}d ${h}h`;
  if (h) return `${h}h ${String(m).padStart(2, "0")}m`;
  return `${m}m`;
}

/** Compact countdown for the menu bar title: "4:49", "0:12". */
export function fmtShort(ms: number): string {
  const mins = Math.max(0, Math.ceil(ms / 60_000));
  return `${Math.floor(mins / 60)}:${String(mins % 60).padStart(2, "0")}`;
}

export function fmtWindow(w: LimitWindow | null, withDate = false, now = Date.now()): string {
  if (!w) return "n/a";
  if (!w.active) return "idle (no session running)";
  if (w.resetsAt !== null && w.resetsAt <= now) return `reset at ${fmtTime(w.resetsAt)}, waiting for next check`;
  const at =
    w.resetsAt === null
      ? "?"
      : withDate
        ? new Date(w.resetsAt).toLocaleString([], { weekday: "short", hour: "2-digit", minute: "2-digit" })
        : fmtTime(w.resetsAt);
  const left = w.resetsAt === null ? "" : ` (in ${fmtDuration(w.resetsAt - now)})`;
  return `${w.usedPct}% used · resets ${at}${left}`;
}

File: src/go/core/format.go
package core

import (
	"fmt"
	"math"
	"time"
)

func FmtTime(ms *int64) string {
	if ms == nil {
		return "?"
	}
	return time.UnixMilli(*ms).Format("3:04 PM")
}

// FmtDuration: "4h 49m", "1d 21h", "12m".
func FmtDuration(d int64) string {
	mins := int64(math.Max(0, math.Round(float64(d)/60_000)))
	days, h, m := mins/1440, (mins%1440)/60, mins%60
	switch {
	case days > 0:
		return fmt.Sprintf("%dd %dh", days, h)
	case h > 0:
		return fmt.Sprintf("%dh %02dm", h, m)
	default:
		return fmt.Sprintf("%dm", m)
	}
}

// FmtShort is the compact countdown for the menu bar title: "4:49", "0:12".
func FmtShort(d int64) string {
	mins := int64(math.Max(0, math.Ceil(float64(d)/60_000)))
	return fmt.Sprintf("%d:%02d", mins/60, mins%60)
}

func FmtWindow(w *LimitWindow, withDate bool, now int64) string {
	if w == nil {
		return "n/a"
	}
	if !w.Active {
		return "idle (no session running)"
	}
	if w.ResetsAt != nil && *w.ResetsAt <= now {
		return fmt.Sprintf("reset at %s, waiting for next check", FmtTime(w.ResetsAt))
	}
	at, left := "?", ""
	if w.ResetsAt != nil {
		at = FmtTime(w.ResetsAt)
		if withDate {
			at = time.UnixMilli(*w.ResetsAt).Format("Mon 3:04 PM")
		}
		left = fmt.Sprintf(" (in %s)", FmtDuration(*w.ResetsAt-now))
	}
	return fmt.Sprintf("%g%% used · resets %s%s", w.UsedPct, at, left)
}
